drivers/media/i2c/isx021.c

```
@@ -0,0 +1,1139 @@
+	ret = isx021_write(sensor, me_addr, val8);
+	if (ret)
+		dev_err(sensor->dev,
+			"[%s] : Failed isx021_write_reg at 0x%x =[ 0x%x]\n",
+			__func__, me_addr, val8);
+
+	return ret;
+}
+
+static u32 isx021_mbus_code(const struct isx021 *sensor)
+{
+	return MEDIA_BUS_FMT_UYVY8_1X16;
+}
+
+static int isx021_power_on(struct isx021 *sensor)
+{
+	int ret;
+
+	ret = regulator_bulk_enable(ARRAY_SIZE(sensor->supplies),
+				    sensor->supplies);
+	if (ret < 0)
+		return ret;
+
+	udelay(1);
+
+	ret = gpiod_direction_output(sensor->reset, 0);
+	if (ret < 0)
+		goto err_supply;
+
+	udelay(1);
+
+	ret = clk_prepare_enable(sensor->clk);
+	if (ret < 0)
+		goto err_reset;
+
+	/*
+	 * The documentation doesn't explicitly say how much time is required
+	 * after providing a clock and before starting I2C communication. It
+	 * mentions a delay of 20µs in 4-wire mode, but tests showed that a
+	 * delay of 100µs resulted in I2C communication failures, while 500µs
+	 * seems to be enough. Be conservative.
+	 */
+	usleep_range(1000, PLUS_10(1000));
+
+	ret = isx021_set_response_mode(sensor);
+	if (ret < 0)
+		goto err_reset;
+
+	return 0;
+
+err_reset:
+	gpiod_direction_output(sensor->reset, 1);
+err_supply:
+	regulator_bulk_disable(ARRAY_SIZE(sensor->supplies), sensor->supplies);
+	return ret;
+}
+
+static void isx021_power_off(struct isx021 *sensor)
+{
+	clk_disable_unprepare(sensor->clk);
+	gpiod_direction_output(sensor->reset, 1);
+	regulator_bulk_disable(ARRAY_SIZE(sensor->supplies), sensor->supplies);
+}
+
+/* ------------------------------------------------------------------------- */
+
+static int isx021_write_mode_set_f_lock_register(struct isx021 *sensor, u8 val8)
+{
+	int ret;
+
+	usleep_range(20000, PLUS_10(20000));
+
+	ret = isx021_write(sensor, 0x8a55, 0x06); // ACK Response mode
+
+	usleep_range(20000, PLUS_10(20000));
+
+	ret = isx021_write(sensor, 0xbef0, val8);
+
+	usleep_range(20000, PLUS_10(20000));
+
+	ret = isx021_write(sensor, 0x8a55, 0x02); // NAK Response mode
```

all errors are being ignored...

---

ACK